demo.cc

```diff
@@ -576,16 +579,134 @@ class Cruncher final {
+    const std::string json = MultiKeyJSON(ev);
+    std::cerr << '@' << demo_id_ << " MixpanelUploader Event: " << json << std::endl;
+    const std::string base64_json = Base64Encode(json);
+    // WORKAROUND(sompylasar): Not using `https://`, could not send HTTPS request.
+    const std::string mixpanel_request = "http://api.mixpanel.com/track?data=" + base64_json;
+    std::cerr << '@' << demo_id_ << " MixpanelUploader Request: " << mixpanel_request << std::endl;
+    if (mixpanel_token_.empty()) {
+      std::cerr << '@' << demo_id_ << " MixpanelUploader Empty token, not sending." << std::endl;
+      return;
+    }
+    auto response = HTTP(GET(mixpanel_request));
+    std::cerr << '@' << demo_id_ << " MixpanelUploader Response: HTTP " << static_cast<int>(response.code) << " \"" << response.body << "\"" << std::endl;
+  }
+
+  template <typename T>
+  inline std::string MultiKeyJSON(T& object) {
+    std::ostringstream os;
+    {
+      // This scope is for `cereal` to flush the archive on scope exit.
+      auto ar =
+          bricks::cerealize::CerealStreamType<bricks::cerealize::CerealFormat::JSON>::CreateOutputArchive(os);
+      // The following allows to make more than one top-level key-value pair.
+      object.serialize(ar);
+    }
+    return os.str();
+  }
+
+  inline std::string Base64Encode(const std::string& str) {
+    // Note: Using `cereal`'s third-party library for base64-encoding.
+    return base64::encode(reinterpret_cast<const unsigned char*>(str.c_str()), str.length());
+  }
+
+  struct MixpanelQuestionAnsweredEvent {
+    struct Properties {
+      // Reserved properties.
+      std::string token;
+      std::string distinct_id;
+      time_t time;
+
+      // Custom properties.
+      schema::QID qid;
+      schema::ANSWER answer;
+
+      template <typename A>
+      void serialize(A& ar) {
+        ar(CEREAL_NVP(token),
+           CEREAL_NVP(distinct_id),
+           CEREAL_NVP(time),
+           cereal::make_nvp("Question", static_cast<size_t>(qid)),
+           cereal::make_nvp("Answer", static_cast<int>(answer)));
+      }
+    };
+
+    std::string event;
+    Properties properties;
+
+    MixpanelQuestionAnsweredEvent(const std::string& token, const schema::AnswerRecord& a) {
+      event = "Question Answered";
+      properties.token = token;
+      properties.distinct_id = a.uid;
+      properties.time = static_cast<uint64_t>(a.ms) / 1000;
+      properties.qid = a.qid;
+      properties.answer = a.answer;
+    }
+
+    template <typename A>
+    void serialize(A& ar) {
+      ar(CEREAL_NVP(event), CEREAL_NVP(properties));
+    }
+  };
+
+ private:
+  const std::string& demo_id_;
+
+  // Note: If `mixpanel_token_` is declared as a reference,
+  // a SIGSEGV happens while in `std::string::assign` at `properties.token = token;`
+  // in `MixpanelQuestionAnsweredEvent` constructor.
+  // TODO(sompylasar) + TODO(dkorolev): Investigate SIGSEGV on assign from a reference.
```

Comment:
W/o looking into the code, I'd guess because it passes in not a temporary string, but a reference to a class member variable, the lifetime of which is all the lifetime of the client with this particular `demo_id_`.